cpp/src/arrow/compute/kernels/scalar_nested.cc

```diff
@@ -819,6 +832,378 @@ const FunctionDoc map_lookup_doc{
+    auto builder = std::make_shared<BuilderType>();
+    ListBuilderType list_builder(ctx->exec_context()->memory_pool(), builder, list_type);
+
+    RETURN_NOT_OK(builder->Reserve(batch.num_values() * batch.length));
+    RETURN_NOT_OK(ReserveBinaryData<InputType>(batch, builder.get()));
+
+    RETURN_NOT_OK(list_builder.Reserve(batch.length));
+
+    for (int i = 0; i < batch.length; ++i) {
+      RETURN_NOT_OK(list_builder.Append());
+      for (const auto& input : batch.values) {
+        if (input.is_array()) {
+          const auto& arr = input.array;
+          if (arr.IsValid(i)) {
+            auto cur_offset = arr.GetValues<OffsetType>(1)[i];
+            auto next_offset = arr.GetValues<OffsetType>(1)[i + 1];
+            std::string_view view(arr.buffers[2].data_as<char>() + cur_offset,
+                                  next_offset - cur_offset);
+            builder->UnsafeAppend(view);
+          } else {
+            builder->UnsafeAppendNull();
+          }
+        } else {
+          builder->UnsafeAppend(UnboxScalar<InputType>::Unbox(*input.scalar));
+        }
+      }
+    }
+    return list_builder.FinishInternal(out->array_data_mutable());
+  }
+
+  // Fixed-size binary types, including decimals
+  template <typename InputType>
+  std::enable_if_t<is_fixed_size_binary_type<InputType>::value, Status> Visit(
+      const InputType& input_type, KernelContext* ctx, const ExecSpan& batch,
+      ExecResult* out) {
+    using BuilderType = typename TypeTraits<InputType>::BuilderType;
+    using ListBuilderType = typename TypeTraits<OutputType>::BuilderType;
+    auto builder = std::make_shared<BuilderType>(input_type.GetSharedPtr(),
+                                                 ctx->exec_context()->memory_pool());
+    ListBuilderType list_builder(ctx->exec_context()->memory_pool(), builder, list_type);
+
+    RETURN_NOT_OK(builder->Reserve(batch.num_values() * batch.length));
+    RETURN_NOT_OK(ReserveBinaryData<InputType>(batch, builder.get()));
+
+    RETURN_NOT_OK(list_builder.Reserve(batch.length));
+
+    for (int i = 0; i < batch.length; ++i) {
+      RETURN_NOT_OK(list_builder.Append());
+      for (const auto& input : batch.values) {
+        if (input.is_array()) {
+          const auto& arr = input.array;
+          if (arr.IsValid(i)) {
+            std::string_view view(arr.buffers[1].data_as<char>() +
+                                      (i + arr.offset) * input_type.byte_width(),
+                                  input_type.byte_width());
+            builder->UnsafeAppend(view);
+          } else {
+            builder->UnsafeAppendNull();
+          }
+        } else {
+          builder->UnsafeAppend(UnboxScalar<InputType>::Unbox(*input.scalar));
+        }
+      }
+    }
+    return list_builder.FinishInternal(out->array_data_mutable());
+  }
+
+  // Deal with nested/union types with a naive approach: First concatenate the inputs,
+  // then shuffle it using Take
```

Comment:
For struct at least, I think there might be a more efficient approach:
* adjoin each child to FIXED_SIZE_LIST, and keep the resulting grandchild array
* create an intermediate struct array for the child arrays computed above 
* create the top level list array using `MakeOffsetsBuffer` and the aforementioned intermediate array

Concretely, if your input type is `struct("a", "b")` and you have the inputs `[{"a": 1, "b": 2}, {"a": 3, "b": 4}]` and `[{"a": 5, "b": 6}, {"a": 7, "b": 8}]`, you would:
1) for field "a", compute `adjoin_as_list([1, 3], [5, 7], output_list_type=FIXED_SIZE_LIST`), giving `[[1, 5], [3, 7]]`
2) keep the child of the latter (ditching its non-existent null bitmap), which is `[1, 5, 3, 7]`
3) same for field "b", giving `[2, 6, 4, 8]`
4) create an intermediate struct array, giving `[{"a": 1, "b": 2}, {"a": 5, "b": 6}, {"a": 3, "b": 4}, {"a": 7, "b": 8}]`
4) compute the top-level offsets `[0, 2, 4]` and combine them with the intermediate nested array, giving the final result `[[{"a": 1, "b": 2}, {"a": 5, "b": 6}], [{"a": 3, "b": 4}, {"a": 7, "b": 8}]]`

That said, this could be left as a TODO with an associated GH issue for someone motivated.

Reply:
Created #37430 and added TODO comment

Reply:
Created #37430 and added TODO comment